lib/foreman_remote_execution_core/script_runner.rb

```diff
@@ -0,0 +1,266 @@
+        channel.request_pty
+        channel.on_data { |ch, data| publish_data(data, 'stdout') }
+        channel.on_extended_data { |ch, type, data| publish_data(data, 'stderr') }
+        # standard exit of the command
+        channel.on_request("exit-status") { |ch, data| publish_exit_status(data.read_long) }
+        # on signal: sedning the signal value (such as 'TERM')
+        channel.on_request("exit-signal") do |ch, data|
+          publish_exit_status(data.read_string)
+          ch.close
+          # wait for the channel to finish so that we know at the end
+          # that the session is inactive
+          ch.wait
+        end
+        channel.exec(command) do |ch, success|
+          @started = true
+          raise("Error initializing command") unless success
+        end
+      end
+      session.process(0) until @started
+      return true
+    end
+
+    def run_sync(command)
+      output = ""
+      exit_status = nil
+      channel = session.open_channel do |ch|
+        ch.on_data { |data| output.concat(data) }
+        ch.on_extended_data { |_, _, data| output.concat(data) }
+        ch.on_request("exit-status") { |_, data| exit_status = data.read_long }
+        # on signal: sedning the signal value (such as 'TERM')
+        ch.on_request("exit-signal") do |_, data|
+          exit_status = data.read_string
+          ch.close
+          ch.wait
+        end
+        ch.exec command do |_, success|
+          raise "could not execute command" unless success
+        end
+      end
+      channel.wait
+      return exit_status, output
+    end
+
+    def su_prefix
+      return if @effective_user.nil? || @effective_user == @ssh_user
+      case @effective_user_method
+      when 'sudo'
+        "sudo -n -u #{@effective_user} "
+      when 'su'
+        "su - #{@effective_user} -c "
+      else
+        raise "effective_user_method ''#{@effective_user_method}'' not supported"
+      end
+    end
+
+    def prepare_known_hosts
+      path = local_command_file('known_hosts')
+      if @host_public_key
+        write_command_file_locally('known_hosts', "#{@host} #{@host_public_key}")
+      end
+      return path
+    end
+
+    def local_command_dir
+      File.join(@local_working_dir, 'foreman-proxy', "foreman-ssh-cmd-#{@id}")
+    end
+
+    def local_command_file(filename)
+      File.join(local_command_dir, filename)
+    end
+
+    def remote_command_dir
+      File.join(@remote_working_dir, "foreman-ssh-cmd-#{id}")
+    end
+
+    def remote_command_file(filename)
+      File.join(remote_command_dir, filename)
+    end
+
+    def ensure_local_directory(path)
+      if File.exist?(path)
+        raise "#{path} expected to be a directory" unless File.directory?(path)
+      else
+        FileUtils.mkdir_p(path)
+      end
+      return path
+    end
+
+    def cp_script_to_remote
+      local_script_file = write_command_file_locally('script', sanitize_script(@script))
+      File.chmod(0777, local_script_file)
```

Comment:
It would be good to integrate the changes proposed here https://github.com/theforeman/smart_proxy_remote_execution_ssh/pull/30, since there won't be much left of rex ssh after this gets merged